### ISSUE TITLE
BBC: Blead Breaks MooX::Press via Type::Tiny

### ISSUE THREAD

**Issue by cjg-cguevara** [Perl/perl5#21006]
This is a bug report for perl from "Carlos Guevara" <carlos@carlosguevara.com>,
generated with the help of perlbug 1.43 running under perl 5.37.11.

-----------------------------------------------------------------

BBC: Blead Breaks MooX::Press

Please see http://fast-matrix.cpantesters.org/?dist=MooX::Press

---
**Flags**
- category=core
- severity=low
---
**Perl configuration**
```
Site configuration information for perl 5.37.11:

Configured by cpan at Fri Apr  7 12:31:50 EDT 2023.

Summary of my perl5 (revision 5 version 37 subversion 11) configuration:
  Commit id: 67d37c55ddffebbcbf4847408106334b9ad9a6bc
  Platform:
    osname=openbsd
    osvers=7.2
    archname=OpenBSD.amd64-openbsd
    uname='openbsd cjg-openbsd7 7.2 generic#7 amd64 '
    config_args='-des -Dprefix=~/bin/perl-blead -Dscriptdir=~/bin/perl-blead/bin -Dusedevel -Duse64bitall'
    hint=recommended
    useposix=true
    d_sigaction=define
    useithreads=undef
    usemultiplicity=undef
    use64bitint=define
    use64bitall=define
    uselongdouble=undef
    usemymalloc=n
    default_inc_excludes_dot=define
  Compiler:
    cc='cc'
    ccflags ='-fno-strict-aliasing -pipe -fstack-protector-strong -I/usr/local/include'
    optimize='-O2'
    cppflags='-fno-strict-aliasing -pipe -fstack-protector-strong -I/usr/local/include'
    ccversion=''
    gccversion='OpenBSD Clang 13.0.0'
    gccosandvers=''
    intsize=4
    longsize=8
    ptrsize=8
    doublesize=8
    byteorder=12345678
    doublekind=3
    d_longlong=define
    longlongsize=8
    d_longdbl=define
    longdblsize=16
    longdblkind=3
    ivtype='long'
    ivsize=8
    nvtype='double'
    nvsize=8
    Off_t='off_t'
    lseeksize=8
    alignbytes=8
    prototype=define
  Linker and Libraries:
    ld='cc'
    ldflags ='-Wl,-E  -fstack-protector-strong -L/usr/local/lib'
    libpth=/usr/lib/clang/13.0.0/lib /usr/lib /usr/local/lib
    libs=-lpthread -lm -lutil -lc
    perllibs=-lpthread -lm -lutil -lc
    libc=/usr/lib/libc.so.96.2
    so=so
    useshrplib=false
    libperl=libperl.a
    gnulibc_version=''
  Dynamic Linking:
    dlsrc=dl_dlopen.xs
    dlext=so
    d_dlsymun=undef
    ccdlflags=' '
    cccdlflags='-DPIC -fPIC '
    lddlflags='-shared -fPIC  -L/usr/local/lib -fstack-protector-strong'


---
@INC for perl 5.37.11:
    /home/cpan/bin/perl-blead/lib/site_perl/5.37.11/OpenBSD.amd64-openbsd
    /home/cpan/bin/perl-blead/lib/site_perl/5.37.11
    /home/cpan/bin/perl-blead/lib/5.37.11/OpenBSD.amd64-openbsd
    /home/cpan/bin/perl-blead/lib/5.37.11

---
Environment for perl 5.37.11:
    HOME=/home/cpan
    LANG (unset)
    LANGUAGE (unset)
    LC_ALL=C
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    PATH=/home/cpan/bin/perl-blead/bin:/home/cpan/bin:/home/cpan/bin:/bin:/sbin:/usr/bin:/usr/sbin:/usr/X11R6/bin:/usr/local/bin:/usr/local/sbin
    PERL_BADLANG (unset)
    SHELL=/usr/local/bin/bash
```

**Comment by jkeenan** [comment]
> This is a bug report for perl from "Carlos Guevara" [carlos@carlosguevara.com](mailto:carlos@carlosguevara.com), generated with the help of perlbug 1.43 running under perl 5.37.11.
> 
> BBC: Blead Breaks MooX::Press
> 
> Please see http://fast-matrix.cpantesters.org/?dist=MooX::Press
I spent *several* hours today trying to identify the cause of the problem in this ticket -- but have come up empty.

I *was* able to reproduce Carlos's test failure on a threaded build of blead (67d37c55ddffebbcbf4847408106334b9ad9a6bc) on FreeBSD-12; see http://www.cpantesters.org/cpan/report/159e185c-d581-11ed-888b-e50daaff8ba7.  However, given the "last good" and "first bad" commit IDs in Carlos's reports, I expected that the "breaking" commit would be very recent, *i.e.,* subsequent to the v5.37.10 release on March 20.  I would not have expected to see these test failures on builds from more than a month ago.  I was therefore surprised that when I attempted to install `MooX::Press` against a build I had lying around from mid-February (d8a109ccd80350f5b83d0928d87032df3ec1b8dd), I got the same test failures; see http://www.cpantesters.org/cpan/report/aaa5e440-d584-11ed-8550-5063aaff8ba7.

I tried bisecting to identify the breaking commit.  I started with this invocation:
```
perl Porting/bisect.pl \
--start=0d292c706462f0a0e7cd64ef002bd8d21513295d \
--end=67d37c55ddffebbcbf4847408106334b9ad9a6bc \
--module=MooX::Press
```
But this invocation ended relatively early with one of those very annoying ambiguous messages:
```
HEAD is now at 692cfdb0a4 Add new release to perlhist
bad - non-zero exit from /tmp/KQTjsTdndw/bin/perl -I /home/jkeenan/.cpan -MCPAN::MyConfig -MCPAN -e $CPAN::Config->{build_dir}=q{/tmp/1ho1gnC6K4}; -e install('MooX::Press'); die unless CPAN::Shell->expand(Module => 'MooX::Press')->uptodate;
Runner returned 256, not 0 for start revision at Porting/bisect.pl line 240.
```
I tried moving the `start` commit earlier, but got similarly unhelpful results with this invocation:
```
perl Porting/bisect.pl \
--start=v5.37.9 \
--end=67d37c55ddffebbcbf4847408106334b9ad9a6bc \
--module=MooX::Press
```
(This was particularly frustrating because this module has a ton of prerequisites and took half an hour to get through only the `end` and `start` commits.)

At this point all I can do is present the test failure output and indicate the CPAN code associated with the breakage.  The failure output for MooX::Press's `t/11plusclass.t` looks like this:
```
t/10plusattr.t ................ ok
package name "::Animal::Dog" is not valid! at /home/jkeenan/testing/d8a109ccd8/lib/perl5/site_perl/5.37.9/Method/Generate/Accessor.pm line 157.
BEGIN failed--compilation aborted at t/11plusclass.t line 47.
t/11plusclass.t ............... 
Dubious, test returned 2 (wstat 512, 0x200)
```
So it's a compile time failure.  `Method::Generate::Accessor` is part of the `Moo` distribution.  The relevant code in that module is:
```
  1 package Method::Generate::Accessor;
  2 use strict;
  3 use warnings;
  4 
  5 use Moo::_Utils qw(_maybe_load_module _install_coderef _module_name_rx);
  6 use Moo::Object ();
  7 BEGIN { our @ISA = qw(Moo::Object) }
  8 use Sub::Quote qw(quote_sub quoted_from_sub quotify sanitize_identifier);
...
 46 sub generate_method {
 47   my ($self, $into, $name, $spec, $quote_opts) = @_;
 48   $quote_opts = {
 49     no_defer => 1,
 50     package => 'Method::Generate::Accessor::_Generated',
 51     %{ $quote_opts||{} },
 52   };
...
150     else {
151       $self->{captures} = {};
152       $methods{$reader} =
153         quote_sub "${into}::${reader}"
154           => '    Carp::croak("'.$reader.' is a read-only accessor") if @_ > 1;'."\n"
155              .$self->_generate_get($name, $spec)
156           => delete $self->{captures}
157           => $quote_opts
158         ;
159     }
```
`Sub::Quote::quote::sub()` looks like this:
```
205 sub quote_sub {
206   # HOLY DWIMMERY, BATMAN!
207   # $name => $code => \%captures => \%options
208   # $name => $code => \%captures
209   # $name => $code
210   # $code => \%captures => \%options
211   # $code
212   my $options =
213     (ref($_[-1]) eq 'HASH' and ref($_[-2]) eq 'HASH')
214       ? pop
215       : {};
216   my $captures = ref($_[-1]) eq 'HASH' ? pop : undef;
217   undef($captures) if $captures && !keys %$captures;
218   my $code = pop;
219   my $name = $_[0];
220   if ($name) {
221     my $subname = $name;
222     my $package = $subname =~ s/(.*)::// ? $1 : caller;
223     $name = join '::', $package, $subname;
224     croak qq{package name "$package" too long!}
225       if length $package > 252;
226     croak qq{package name "$package" is not valid!}
227       unless $package =~ /^[^\d\W]\w*(?:::\w+)*$/;
```
Can anyone construct a reduced case?

Thank you very much.
Jim Keenan

**Comment by andk** [comment]
It's a Type-Tiny thingy. I just picked an arbitrary perl 5.37.4 with a tested MooX::Press (report: http://www.cpantesters.org/cpan/report/7d4d1390-a6d3-11ed-b19e-dcb3eb881259). I tested TOBYINK/MooX-Press-0.086.tar.gz again with this perl and it was still passing its test. I then upgraded Type-Tiny to TOBYINK/Type-Tiny-2.004000.tar.gz and tested MooX-Press another time. Now it failed (report: http://www.cpantesters.org/cpan/report/0dc0d39e-d5dc-11ed-9345-d454025c05f7)

@tobyink please check

**Comment by jkeenan** [comment]
> 

Confirmed.

**Comment by jkeenan** [comment]
My debugging today identified a commit to the Type::Tiny codebase at which test t/11plusclass.t in MooX::Press began to fail.
```
commit 99c857df4a56194a3142bb00af4d5a8b2d23148d
Author:     Toby Inkster <mail@tobyinkster.co.uk>
AuthorDate: Sat Apr 1 18:01:26 2023 +0100
Commit:     Toby Inkster <mail@tobyinkster.co.uk>
CommitDate: Sat Apr 1 18:01:26 2023 +0100

    Inherit exception_class from parent type when possible
```
https://github.com/tobyink/p5-type-tiny/commit/99c857df4a56194a3142bb00af4d5a8b2d23148d

This has been reported in https://github.com/tobyink/p5-moox-press/issues/4.

I installed perl-5.37.10, installed an older version of Type-Tiny (2.002000) from CPAN; installed latest Sub-Quote and MooX-Press.  Then did a git checkout of Type-Tiny from GH.  Started from tag 2.002000, moved forward until just before commits whose messages referred to exceptions.  Built and installed Type::Tiny at that commit. Re-ran the failing test in MooX::Press.  It PASSed. Moved forward one commit to that listed above.  Rebuilt and reinstalled Type::Tiny.  Re-ran the test in MooX::Press which now FAILed.

I believe that this demonstrates that the test failure reported in MooX::Press in this ticket was not triggered by a change in Perl 5 blead.  I am taking this ticket for the purpose of closing it in a few days unless someone objects.
